### ISSUE TITLE
Add zap implementation of logger in contrib

### ISSUE THREAD

**Issue by cretz** [temporalio/sdk-go#704]
**Is your feature request related to a problem? Please describe.**

Lots of people are manually implementing `log.Logger` using Zap.

**Describe the solution you'd like**

* Add a `zap` child module to the `contrib` folder.
* It should rely on the `*zap.S`'s `<Level>W` calls which are higher performing than you see from other people's adapters
* Fix up https://github.com/temporalio/samples-go/tree/main/zapadapter to use it
* Update the vanity URL at `go.temporal.io/sdk/contrib/zap` and tag a release for `contrib/zap/v0.1.0`

**Comment by ianthpun** [comment]
> It should rely on the *zap.S's <Level>W calls which are higher performing than you see from other people's adapters

shouldn't the warning level calls just be mapped to the `Warn` call that satisfies the `log.Logger` interface and not to any other calls? 
https://github.com/temporalio/samples-go/blob/main/zapadapter/zap_adapter.go#L71-L73

`Warn` would call zaploggers `Warn` while `Info` should call `Info`, etc

**Comment by cretz** [comment]
Correct. When I said "It should rely on the *zap.S's W", what I mean is it should use `zap.SugaredLogger.<level>W` calls which are equivalent to our key-value varargs.

**Comment by cretz** [comment]
Note, `slog` is now coming, see #1158
